File: ios/EstudioReformer/Networking/AppSession.swift
import Foundation
import SwiftUI

/// Root auth/session state — who's logged in, their approval status, and the
/// shared `APIClient` every screen talks through. Token persistence here is
/// UserDefaults for demo simplicity; swap for Keychain before shipping.
@MainActor
final class AppSession: ObservableObject {
    @Published var user: PublicUser?
    @Published var isLoadingInitialAuth = true
    @Published var lastError: String?

    let api = APIClient()

    private let tokenKey = "auth_token"

    init() {
        api.token = UserDefaults.standard.string(forKey: tokenKey)
        Task { await bootstrap() }
    }

    private func bootstrap() async {
        defer { isLoadingInitialAuth = false }
        guard api.token != nil else { return }
        do {
            let me: MeResponse = try await api.request("/auth/me")
            user = me.user
        } catch {
            // Stored token is stale/invalid — fall back to logged-out.
            signOut()
        }
    }

    func register(name: String, email: String, phone: String, password: String,
                  experienciaPrevia: Bool, patologia: Bool, patologiaTexto: String) async throws {
        struct Body: Encodable {
            let name: String, email: String, phone: String, password: String
            let experienciaPrevia: Bool, patologia: Bool, patologiaTexto: String
        }
        let res: AuthResponse = try await api.request(
            "/auth/register", method: "POST",
            body: Body(name: name, email: email, phone: phone, password: password,
                       experienciaPrevia: experienciaPrevia, patologia: patologia, patologiaTexto: patologiaTexto)
        )
        persist(res)
    }

    func login(email: String, password: String) async throws {
        struct Body: Encodable { let email: String; let password: String }
        let res: AuthResponse = try await api.request("/auth/login", method: "POST", body: Body(email: email, password: password))
        persist(res)
    }

    func refreshMe() async {
        guard api.token != nil else { return }
        if let me: MeResponse = try? await api.request("/auth/me") {
            user = me.user
        }
    }

    func signOut() {
        api.token = nil
        user = nil
        UserDefaults.standard.removeObject(forKey: tokenKey)
    }

    private func persist(_ res: AuthResponse) {
        api.token = res.token
        user = res.user
        UserDefaults.standard.set(res.token, forKey: tokenKey)
    }
}

File: android/app/src/main/java/es/estudioreformer/app/data/AppSession.kt
package es.estudioreformer.app.data

import android.app.Application
import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import es.estudioreformer.app.network.ApiClient
import es.estudioreformer.app.network.LoginBody
import es.estudioreformer.app.network.PublicUser
import es.estudioreformer.app.network.RegisterBody
import kotlinx.coroutines.launch

/** Root auth/session state — who's logged in, their approval status, and the
 * shared [ApiClient] every screen talks through. Token persistence here is
 * plain SharedPreferences for demo simplicity; swap for EncryptedSharedPreferences
 * or the Keystore before shipping. */
class AppSession(application: Application) : AndroidViewModel(application) {
    val api = ApiClient()

    var user by mutableStateOf<PublicUser?>(null)
        private set
    var isLoadingInitialAuth by mutableStateOf(true)
        private set

    private val prefs = application.getSharedPreferences("auth", Context.MODE_PRIVATE)

    init {
        api.token = prefs.getString("token", null)
        viewModelScope.launch {
            if (api.token != null) {
                try {
                    user = api.me().user
                } catch (e: Exception) {
                    signOut()
                }
            }
            isLoadingInitialAuth = false
        }
    }

    suspend fun register(
        name: String, email: String, phone: String, password: String,
        experienciaPrevia: Boolean, patologia: Boolean, patologiaTexto: String,
    ) {
        val res = api.register(RegisterBody(name, email, phone, password, experienciaPrevia, patologia, patologiaTexto))
        persist(res.token, res.user)
    }

    suspend fun login(email: String, password: String) {
        val res = api.login(LoginBody(email, password))
        persist(res.token, res.user)
    }

    suspend fun refreshMe() {
        if (api.token == null) return
        try { user = api.me().user } catch (e: Exception) { /* keep last known state */ }
    }

    fun signOut() {
        api.token = null
        user = null
        prefs.edit().remove("token").apply()
    }

    private fun persist(token: String, u: PublicUser) {
        api.token = token
        user = u
        prefs.edit().putString("token", token).apply()
    }
}
